content/developers/developer-patterns/navigating-merklelogs/index.md

```diff
@@ -0,0 +1,748 @@
+
+## The trieData entries are 512 bits each and are formed from two fields
+
+With the trie keys, full data recovery can be verified without needing to remember the order that events were added to the log.
+
+Massifs can be recovered independently even if the data for other massifs is incomplete.
+
+The index also provides for proof of exclusion. Proof that a specific event is not in the index. As the event identities contribute to both the index keys and the leaf entries, and as the keys are time ordered on addition, it is possible to audit the specific range.
+This will confirm the log is consistent with the index for the time range in which the disputed item is claimed to exist.
+
+The trieData section is 2 \* 32 \* 2<sup>height</sup> bytes long, which is exactly double what is needed.
+For a standard massif height of 14, it has 8,192 entries in the first 524,288 bytes.
+The subsequent 524,288 will always be zero.
+The format of each entry is then, for a massif height of 14:
+
+```output
++----------------+
+| HEADER DATA    |
++----------------+
+
+288
++----------------+
+| TRIE-DATA      |
+|                |
+|                |
+|................| 288 + 524288 = 524576
+|  always        |
+|                |
+|       zero     |
++----------------+ 288 + 524288 * 2 = 1048864
+```
+
+Each entry is formatted like this
+
+```output
+0                                                        31
+SHA256(BYTE(0x00) || tenant-identity || event.identity)
+0                                       BYTES(IDTIMESTAMP)
+32                                      56               63
+```
+
+Described in further detail in the [term-cheatsheet](https://github.com/datatrails/go-datatrails-merklelog/blob/main/term-cheatsheet.md#trie-entry)
+Note that the idtimestamp is unique to the tenant and the DataTrails service.
+When sharing events with other tenants, the idtimestamp will not correlate directly with activity in other tenant logs.
+
+When a massif is initialized the trieData is pre-populated for all leaves and set to all zero bytes.
+As events are recorded in the log, the zero-padded index is filled in.
+A sub-range of field 0 will change when saving the last idtimestamp in it.
+The MMR node values are strictly only ever appended to the blob.
+Once appended they will never change and they will never move.
+
+Returning to the trieData section, given an event identity of:  
+`assets/87dd2e5a-42b4-49a5-8693-97f40a5af7f8/events/a022f458-8e55-4d63-a200-4172a42fc2a`
+
+The trieKey would be:  
+`eda55087407b2e6f52c668f039c624d2382422ea8c765d618a0bbbef2936223d`
+
+The following python snippet generates a trieKey from the event data to confirm what should be in the index at a specific position.
+
+{{< tabs >}}
+   {{< tab name="Python" >}}
+
+   ```python
+  def triekey(tenant, event):
+      h = hashlib.sha256()
+      h.update(bytes([0]))
+      h.update(tenant.encode())
+      h.update(event.encode())
+      return h.hexdigest()
+
+  print(triekey(
```

Comment:
Need to decide how we'll switch betwen python, bash to curl, then back to python